features/workflow.feature

```diff
@@ -58,3 +64,71 @@ Feature: Workflow
         And the "BAZ" build succeeded
         Then the "XYZZY" job is started
         And that "XYZZY" build uses the same SHA as the "FOO" build
+
+    Scenario: Branch filtering (the pipeline's branch is committed)
+        Given an existing pipeline with the workflow:
+            | job   | requires        |
+            | FOO   | ~commit         |
+            | BAR   | ~commit:staging |
+            | BAZ   | ~commit:/^.*$/  |
+        When a new commit is pushed
+        And it is against the pipeline's branch
+        Then the "FOO" job is started
+        And the "BAR" job is not started
+        And the "BAZ" job is started
+
+    Scenario: Branch filtering (the staging branch is committed)
+        Given an existing pipeline with the workflow:
+            | job   | requires        |
+            | FOO   | ~commit         |
+            | BAR   | ~commit:staging |
+            | BAZ   | ~commit:/^.*$/  |
+        When a new commit is pushed
+        And it is against the staging branch
+        Then the "FOO" job is not started
+        And the "BAR" job is started
+        And the "BAZ" job is started
+
+    Scenario: Branch filtering (a pull request is opened to the pipeline's branch)
+        Given an existing pipeline with the workflow:
+            | job   | requires        |
+            | FOO   | ~pr             |
+            | BAR   | ~pr:staging     |
+            | BAZ   | ~pr:/^.*$/      |
+            | QUX   | ~pr:/master     |
+        When a pull request is opened
+        And it is targeting the pipeline's branch
+        Then the "FOO" job is started
+        And the "BAR" job is not started
+        And the "BAZ" job is started
+        And the "QUX" job is started
+
+    Scenario: Branch filtering (pr and filtered pr workflow)
+        Given an existing pipeline with the workflow:
+            | job   | requires        |
+            | FOO   | ~pr             |
+            | FOO2  | FOO             |
+            | QUX   | ~pr:/master     |
+            | QUX2  | QUX             |
+        When a pull request is opened
+        And it is targeting the pipeline's branch
+        Then the "FOO" job is started
+        And the "QUX" job is started
+        Then the "FOO" job is succeeded
+        And the "QUX" job is succeeded
+        Then the "FOO2" job is not started
```

Comment:
should FOO2 be started?

Reply:
This example is a little strange but `FOO2` is not started. `FOO2` job requires `FOO` and `FOO` requires `~pr`.  `~pr` will not trigger subsequence jobs.